File: hashmap/146.lru-cache.py
#
# @lc app=leetcode id=146 lang=python3
#
# [146] LRU Cache
#

# @lc code=start


class DLinkNode:
    def __init__(self, key, val):
        self.key = key
        self.val = val
        self.prev = None
        self.next = None


class LRUCache:
    def __init__(self, capacity: int):
        self.cap = capacity
        self.map = {}
        self.top = DLinkNode(None, -1)
        self.tail = DLinkNode(None, -1)
        self.top.next = self.tail
        self.tail.prev = self.top

    def get(self, key: int) -> int:
        if key in self.map:
            cur = self.map[key]
            # move cur node
            cur.next.prev = cur.prev
            cur.prev.next = cur.next
            # move cur to first
            top = self.top.next
            self.top.next = cur
            cur.prev = self.top
            cur.next = top
            top.prev = cur

            return self.map[key].val
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.map:
            print("key: ", key)
            cur = self.map[key]
            cur.val = value
            # move cur node
            cur.next.prev = cur.prev
            cur.prev.next = cur.next
            # move cur to first
            top = self.top.next
            self.top.next = cur
            cur.prev = self.top
            cur.next = top
            top.prev = cur
        else:
            cur = DLinkNode(key, value)
            self.map[key] = cur
            top = self.top.next
            self.top.next = cur
            cur.prev = self.top
            cur.next = top
            top.prev = cur
            if len(self.map.keys()) > self.cap:
                self.map.pop(self.tail.prev.key)
                # remove tail Node
                self.tail.prev.prev.next = self.tail
                self.tail.prev = self.tail.prev.prev

    def __repr__(self):
        vals = []
        p = self.top.next
        while p.next:
            vals.append(str(p.val))
            p = p.next
        return "->".join(vals)


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end

if __name__ == "__main__":
    #     ["LRUCache","put","put","get","put","put","get"]
    # ' +
    #   '[[2],[2,1],[2,2],[2],[1,1],[4,1],[2]]
    # Answer
    # [null,null,null,1,null,null,-1]
    # Expected Answer
    # [null,null,null,2,null,null,-1]
    lru = LRUCache(2)
    lru.put(2, 1)
    lru.put(2, 2)
    lru.get(2)
    lru.put(1, 1)
    lru.put(4, 1)
    lru.get(2)
    print(lru)

File: hashmap/146.lru-cache.go
package main

import "container/list"

/*
 * @lc app=leetcode id=146 lang=golang
 *
 * [146] LRU Cache
 */

// @lc code=start
// type dlistNode struct {
//     prev  *dlistNode
//     next  *dlistNode
//     key   int
//     val   int
//     hnext *dlistNode
// }

type LRUCache struct {
	cap int
	l   *list.List
	m   map[int]*list.Element
}

type Pair struct {
	key   int
	value int
}

func Constructor(capacity int) LRUCache {
	return LRUCache{
		cap: capacity,
		l:   new(list.List),
		m:   make(map[int]*list.Element, capacity),
	}
}

func (this *LRUCache) Get(key int) int {
	if node, ok := this.m[key]; ok {
		val := node.Value.(*list.Element).Value.(Pair).value
		this.l.MoveToFront(node)
		return val
	}
	return -1
}

func (this *LRUCache) Put(key int, value int) {
	if node, ok := this.m[key]; ok {
		// val := node.Value.(*list.Element).Value.(Pair).value
		this.l.MoveToFront(node)
		node.Value.(*list.Element).Value = Pair{key: key, value: value}
	} else {
		if this.l.Len() == this.cap {
			idx := this.l.Back().Value.(*list.Element).Value.(Pair).key
			delete(this.m, idx)
			this.l.Remove(this.l.Back())
		}
		node := &list.Element{
			Value: Pair{
				key:   key,
				value: value,
			},
		}
		ptr := this.l.PushFront(node)
		this.m[key] = ptr
	}
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * obj := Constructor(capacity);
 * param_1 := obj.Get(key);
 * obj.Put(key,value);
 */
// @lc code=end
